### DISCUSSION TITLE
Creating custom drivers module in the `zephyrproject-rtos/example-application`

### DISCUSSION THREAD

**Discussion by IvanVnucec** [zephyrproject-rtos/zephyr#56932] [Q&A] [answered]
I'm developing custom application with the [example-application](https://github.com/zephyrproject-rtos/example-application), and now I want to create a new driver out-of-tree module called latching_relay with its own custom API. Next, I will create the driver for the latching relay part (some Panasonic one) that implements the latching_relay API. I've read about system drivers [here](https://docs.zephyrproject.org/3.2.0/kernel/drivers/index.html#system-drivers) but didn't find any useful information regarding creating custom driver model in the out-of-tree application.
I've copied the `hello_world_module` from the `out_of_tree_driver` sample located in the Zephyr `samples` folder, successfully build it and flashed, and its working. Next, I would like to move it to the drivers folder of the example application. When I move `hello_world_module` folder to the `drivers` folder and edit `app/CMakeLists.txt` file to append zephyr extra modules with the path `${CMAKE_CURRENT_SOURCE_DIR}/../drivers/hello_world_module`, I get a fatal error saying `No such file or directory #include <syscalls/hello_world_driver.h>`. I have `CONFIG_APPLICATION_DEFINED_SYSCALL` flag enabled in the `prj.conf`.

Because my relay is driven by simply changing the states of two GPIO's, I could implement the relay driver as LED-GPIO one but I don't have good feeling doing so because my relay is not an LED.

**Comment by yashi** [comment, marked as answer]
Please help us to help you by providing a [minimal working example](https://en.wikipedia.org/wiki/Minimal_reproducible_example).

### TL;DR

Use the following `CMakeLists.txt` instead of the original.
```cmake
# SPDX-License-Identifier: Apache-2.0

if(CONFIG_HELLO_WORLD_DRIVER)
  # Add hello_world_driver.h to the set of global include paths.
  zephyr_include_directories(.)

  list(APPEND SYSCALL_INCLUDE_DIRS ${ZEPHYR_CURRENT_MODULE_DIR})
  set(SYSCALL_INCLUDE_DIRS ${SYSCALL_INCLUDE_DIRS} PARENT_SCOPE)

  zephyr_library()
  zephyr_library_sources(
    hello_world_driver.c
    )
endif()
```

### Details

The reason `syscalls/hello_world_driver.h` is not generated is that the build system doesn't know where the `hello_world_driver.h` is.

When you build `samples/application_development/out_of_tree_driver`, `hello_world_driver.h` is under the application directory, which is `samples/application_development/out_of_tree_driver`.  The build system scans the application directory because of  `CONFIG_APPLICATION_DEFINED_SYSCALL=y`.

When you _move_ `hello_world_module` out from the application directory, even though the build sytem scans the application directory, it can't find `hello_world_driver.h` for the obvious reason.

Instead of `CONFIG_APPLICATION_DEFINED_SYSCALL=y`, you need to tell where to find the header file.  To do so, you can add the following two lines in the module's CMakeLists.txt.

```cmake
  list(APPEND SYSCALL_INCLUDE_DIRS ${ZEPHYR_CURRENT_MODULE_DIR})
  set(SYSCALL_INCLUDE_DIRS ${SYSCALL_INCLUDE_DIRS} PARENT_SCOPE)
```

These lines are from the last part from the section "[Build system integration](https://docs.zephyrproject.org/3.1.0/develop/modules.html#build-system-integration)":

> It is possible to append values to a Zephyr CMake list variable from the module’s first CMakeLists.txt file. To do so, append the value to the list and then set the list in the PARENT_SCOPE of the CMakeLists.txt file. For example, to append bar to the FOO_LIST variable in the Zephyr CMakeLists.txt scope:
> 
> ```cmake
> list(APPEND FOO_LIST bar)
> set(FOO_LIST ${FOO_LIST} PARENT_SCOPE)
> ```
> 
> An example of a Zephyr list where this is useful is when adding additional directories to the `SYSCALL_INCLUDE_DIRS` list.

**Reply by IvanVnucec** [reply]
> Please help us to help you by providing a [minimal working example](https://en.wikipedia.org/wiki/Minimal_reproducible_example).
> ### TL;DR
> 
> Use the following `CMakeLists.txt` instead of the original.
> 
> ```cmake
> # SPDX-License-Identifier: Apache-2.0
> 
> if(CONFIG_HELLO_WORLD_DRIVER)
>   # Add hello_world_driver.h to the set of global include paths.
>   zephyr_include_directories(.)
> 
>   list(APPEND SYSCALL_INCLUDE_DIRS ${ZEPHYR_CURRENT_MODULE_DIR})
>   set(SYSCALL_INCLUDE_DIRS ${SYSCALL_INCLUDE_DIRS} PARENT_SCOPE)
> 
>   zephyr_library()
>   zephyr_library_sources(
>     hello_world_driver.c
>     )
> endif()
> ```
> 
> ### Details
> 
> The reason `syscalls/hello_world_driver.h` is not generated is that the build system doesn't know where the `hello_world_driver.h` is.
> 
> When you build `samples/application_development/out_of_tree_driver`, `hello_world_driver.h` is under the application directory, which is `samples/application_development/out_of_tree_driver`. The build system scans the application directory because of `CONFIG_APPLICATION_DEFINED_SYSCALL=y`.
> 
> When you _move_ `hello_world_module` out from the application directory, even though the build sytem scans the application directory, it can't find `hello_world_driver.h` for the obvious reason.
> 
> Instead of `CONFIG_APPLICATION_DEFINED_SYSCALL=y`, you need to tell where to find the header file. To do so, you can add the following two lines in the module's CMakeLists.txt.
> 
> ```cmake
>   list(APPEND SYSCALL_INCLUDE_DIRS ${ZEPHYR_CURRENT_MODULE_DIR})
>   set(SYSCALL_INCLUDE_DIRS ${SYSCALL_INCLUDE_DIRS} PARENT_SCOPE)
> ```
> 
> These lines are from the last part from the section "[Build system integration](https://docs.zephyrproject.org/3.1.0/develop/modules.html#build-system-integration)":
> 
> > It is possible to append values to a Zephyr CMake list variable from the module’s first CMakeLists.txt file. To do so, append the value to the list and then set the list in the PARENT_SCOPE of the CMakeLists.txt file. For example, to append bar to the FOO_LIST variable in the Zephyr CMakeLists.txt scope:
> > ```cmake
> > list(APPEND FOO_LIST bar)
> > set(FOO_LIST ${FOO_LIST} PARENT_SCOPE)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > An example of a Zephyr list where this is useful is when adding additional directories to the `SYSCALL_INCLUDE_DIRS` list.

Thank you for your explanation. I've managed to create custom drivers module with the syscalls enabled with the help of the [teslabs/zds-2022-drivers-app ](https://github.com/teslabs/zds-2022-drivers-app/) repository where you can find the appending to the `SYSCALL_INCLUDE_DIRS` variable as you've mentioned. Thanks again.
